Introduce secp256k1 module with field and group classes to test framework

sipa opened the pull request:
This PR rewrites a portion of `test_framework/key.py`, in a compatible way, by introducing classes that encapsulate field element and group element logic, in an attempt to be more readable and reusable.

To maximize readability, the group element logic does not use Jacobian coordinates. Instead, group elements just store (affine) X and Y coordinates directly. To compensate for the performance loss this causes, field elements are represented as fractions. This undoes most, but not all, of the performance loss, and there is a few % slowdown (as measured in `feature_taproot.py`, which heavily uses this).

The upside is that the implementation for group laws (point doubling, addition, subtraction, ...) is very close to the mathematical description of elliptic curves, and this extends to potential future extensions (e.g. ElligatorSwift as needed by #27479).

sipa committed:
test: add secp256k1 module with FE (field element) and GE (group element) classes

These are primarily designed for ease of understanding, not performance.

amovfx commented on test/functional/test_framework/key.py:
```diff
@@ -25,258 +25,329 @@ def TaggedHash(tag, data):
+        assert a is None
+        return self
+
+    def __sub__(self, a):
+        """Subtract two points, or subtract infinity from a point."""
+        if a is None:
+            return self
+        return self + (-a)
```
I'm curious as to why the return value isnt self - a?
is it because this function calls the previously declared  `__add__ ` and requires a negative input to evaluate properly?
or (-a) is calling the `__mul__`

amovfx reviewed (commented):
Reviewed to learn.

theStack committed:
test: EC: optimize scalar multiplication of G by using lookup table

On my machine, this speeds up the functional test feature_taproot.py by
a factor of >1.66x (runtime decrease from 1m16.587s to 45.334s).

Co-authored-by: Pieter Wuille <pieter@wuille.net>

theStack reviewed (commented):
Concept ACK, +1 on increased readability and reusability :100: (found my way here through #24748)

The slow-down is unfortunately pretty significant on the two x86_64 machines I tested, `feature_taproot.py` taking almost twice as long compared to master here. Results measured via `time`:

| Machine / OS  | master      |  pr26222 (rebased) |
|----------|:-------------:|:------:|
| AMD EPYC 7702P (6) / Ubuntu 22.04.1 LTS |  40.496s | 1m16.587s |
| Intel (Broadwell) (2) / OpenBSD 7.2 |    2m57.43s   |   5m45.90s |

For the fun of it, I looked a bit into other ways how to compensate the performance loss. Considering that scalar multiplication with G is a frequent operation, precomputing G's doubled-up values (G, 2G, 4G, ..., (2^254)G, (2^255)G) in a lookup table seemed promising, with the goal of needing only ~128 point additions on average (one for each bit set in the scalar), without any extra double operations. Using that, the execution time is still not matching, but at least pretty close to the master branch (45.334s for first, 3m23.48s for second machine), without too much additional code (~25 LOC):

<details>
<summary>diff for commit "test: EC: optimize scalar multiplication of G by using lookup table"</summary>

```diff
index 7434ba49e..1ad1bc3e4 100644
--- a/test/functional/test_framework/key.py
+++ b/test/functional/test_framework/key.py
@@ -233,6 +233,8 @@ class GE:
 
     def __mul__(self, a):
         """Multiply a point with an integer (scalar multiplication)."""
+        if self == SECP256K1_G:  # optimize generator multiplication using precomputed data
+            return fast_g.mul(a)
         r = None
         for i in range(a.bit_length() - 1, -1, -1):
             if r is not None:
@@ -596,6 +598,31 @@ def sign_schnorr(key, msg, aux=None, flip_p=False, flip_r=False):
     e = int.from_bytes(TaggedHash("BIP0340/challenge", R.to_bytes_xonly() + P.to_bytes_xonly() + msg), 'big') % GE.ORDER
     return R.to_bytes_xonly() + ((k + e * sec) % GE.ORDER).to_bytes(32, 'big')
 
+
+class FastG:
+    """Speed up scalar multiplication with the generator point G
+       by using a precomputed lookup table with its powers of 2:
+       g_table = [G, G*2, G*4, G*(2^3), G*(2^4), ..., G*(2^255)]
+       The points corresponding to each bit set in the scalar are
+       added up, i.e. on average ~128 point additions take place.
+    """
+    def __init__(self):
+        self.g_table = []  # g_table[i] = G * (2^i)
+        g_i = SECP256K1_G
+        for bit in range(256):
+            self.g_table.append(g_i)
+            g_i = g_i.double()
+
+    def mul(self, a):
+        result = None
+        for bit in range(a.bit_length()):
+            if (a & (1 << bit)):
+                result += self.g_table[bit]
+        return result
+
+fast_g = FastG()
+
+
 class TestFrameworkKey(unittest.TestCase):
     def test_schnorr(self):
         """Test the Python Schnorr implementation."""
```
</details>

(see branch https://github.com/theStack/bitcoin/tree/pr26222_followup_precompute_g_doubles)

Not 100% sure if we even want optimizations like this in test framwork (also, immediately generating data at module import seems kind of hacky), but it may be worth it in this case? I'm pretty sure there are more efficient ways to pursue this idea (e.g. extending the lookup table for larger chunks than only 1-bit pieces to need even less additions), but probably the extra implementation / code clutter / maintenance effort is not worth it. Any other ideas? Happy to hear shadowy secp256k1 super-magicians' thoughts here :)

fanquake commented:
cc @real-or-random

hebasto reviewed (commented):
> To maximize readability...

Concept ACK on that. At least, non-cryptography abstract algebra background is enough to review the code :)

Partially reviewed. The `class FE` looks good.

pinheadmz reviewed (commented):
concept ACK

Did as much code review as my tiny brain could handle. I agree that this implementation is easy to read and follow. I checked as much as I could against [NIST params](https://www.secg.org/sec2-v2.pdf) and equations from [Guide to Elliptic Curve Cryptography](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.394.3037&rep=rep1&type=pdf)

Comparing run time of `feature_taproot.py` between master and branch, over a few trials with and without RAM disk, the new code only cost me about 5 seconds or about 5% of total runtime.

instagibbs commented on test/functional/test_framework/key.py:
```diff
@@ -23,258 +23,350 @@ def TaggedHash(tag, data):
+
+    def __str__(self):
+        """Convert this group element to a string."""
+        return f"({self.x},{self.y})"
```
it's defined by `FE` as hex, so should be fine?

instagibbs commented:
Reads fine to me from my not-even-cryptographer-on-tv level of understanding.

Goes from 27 to 31 seconds on my machine with the precomputed table, which is very easy to understand.

real-or-random reviewed (commented):
Concept ACK on making the code simple 

Since this is something like a reference implementation, I think `GE` should get the infinity handling right.  If you agree that's a good idea, then it's perhaps better to have a more explicit representation for infinity than `None`.

real-or-random commented on test/functional/test_framework/key.py:
```diff
@@ -23,258 +23,349 @@ def TaggedHash(tag, data):
+class GE:
+    """Objects of this class represent points (group elements) on the secp256k1 curve.
+
+    The point at infinity is represented as None."""
```
outdated comment

real-or-random commented on test/functional/test_framework/key.py:
```diff
@@ -23,258 +23,349 @@ def TaggedHash(tag, data):
+        return GE(x3, y3)
+
+    def __add__(self, a):
+        """Add two points, or a point and infinity, together."""
```
This one too. I suggest iterating over the all doc comments, also to make "group element" vs "point" consistent. (`__repr__` uses group element.)

real-or-random commented on test/functional/test_framework/secp256k1.py:
Perhaps the comment should be explicit about the behavior when `v >= FE.SIZE`

theStack commented on test/functional/test_framework/secp256k1.py:
Regarding the goal of simple implementation, is there a strong need to keep the `FE.is_square` method? Right now this is the only call-site, which I think could be replaced by
```suggestion
        return (FE(x)**3 + 7).sqrt() is not None
```
I assume trying to actually square is way slower than computing the Jacobi symbol, but since `is_valid_x` doesn't appear to be in a critical code-path, that's probably fine? At least for `feature_taproot.py` I didn't see a decrease in performance.

theStack commented on test/functional/test_framework/secp256k1.py:
nit: like `from_bytes_xonly` does, could throw an assertion error here if the size of the passed encoding is invalid (if a test writer e.g. accidentally passes an x-only-pubkey to `from_bytes`, getting an assertion is likely more helpful than only getting `None`):
```suggestion
        assert len(b) in (33, 65)
        if len(b) == 33:
```
(or alternatively, add an `else:` branch below with `assert False, "size must be 33 or 65"` or something alike)

theStack reviewed (approved):
ACK ab30e81b3bb9e9b02ca1ac835eef04efdf87b4e1

theStack reviewed (approved):
re-ACK d4fb58ae8ae9772d025ead184ef8f2c0ea50df3e

(as per `$ git range-diff ab30e81b...d4fb58ae`, verified that the only changes since the last ACK were a rebase on master and https://github.com/bitcoin/bitcoin/pull/26222#discussion_r1235947681 tackled)

stratospher reviewed (approved):
tested ACK d4fb58a. really liked how this PR makes the secp256k1 code in the tests more intuitive and easier to follow!

achow101 commented:
ACK d4fb58ae8ae9772d025ead184ef8f2c0ea50df3e

Much simpler to understand, and the more complicated stuff matches up with the descriptions in WIkipedia.
